### PR TITLE
feat!: introduce split definitions

### LINKED ISSUE #120: Provide JSON Schema draft 7 file as part of schemas
#### Reason/Context
[JSON Schema Draft 7](https://json-schema.org/draft-07/json-schema-release-notes.html) is the JSON Schema version AsyncAPI is compatible with. 
All our schemas are written with such dialect, ergo we specify it through the [$schema](https://json-schema.org/understanding-json-schema/reference/schema.html#id4) field.
We also make references to that same schema when referencing through `$ref`.

As you can notice, we strongly depend on such JSON Schema schema. 
I'm suggesting we include it as part of our schemas collection so any tool can directly load all schemas, including the JSON Schema Draft 7 on their Schema registries so they don't need to fetch it externally.

A recent real use case can be found https://github.com/asyncapi/studio/pull/159, where the JSON Schema Draft 7 should be provided to Monaco editor YAML plugin in order to avoid fetching it through a HTTP Request.

The plan will also include one JSON Schema schema file per version we require, meaning if we ever adopt a newer JSON Schema draft, we will include it as well here.

#### Description
Add a new file into [/schemas](https://github.com/asyncapi/spec-json-schemas/tree/master/schemas) called `json-schema-draft-07.json` with the exact content of http://json-schema.org/draft-07/schema.

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR introduces split definitions and bundling behavior to provide better maintainability of the JSON Schema files.

Changes:
-  Specification extension pattern property regex needs to be valid against u flag. This means that `^x-[\\w\\d\\.\\-\\_]+$` is changed to `^x-[\\w\\d.\\-_]+$`.
-  Specification components pattern property regex needs to be valid against u flag. This means that `^x-[\\w\\d\\.\\-\\_]+$` is changed to `^x-[\\w\\d.\\-_]+$`
-  Pattern property regex for parameter and correlationId location needs to be valid against u flag. This means that `^\\$message\\.(header|payload)\\#(\\/(([^\\/~])|(~[01]))*)*` is changed to `^\\$message\\.(header|payload)#(\\/(([^\\/~])|(~[01]))*)*`
-  Fixes $ref is incorrectly defined for a server.
- All schema id's `$id` now follow `http://asyncapi.com/definitions/<version>/<name>.json` instead of `<name>`. The reason for this change is to follow an accurate bundling behavior from JSON Schema.

As this is a breaking change you can read the [migration guide](https://github.com/asyncapi/spec-json-schemas/blob/master/migrations/Migrate%20to%20version%203.md) to get more information.

**Related issue(s)**
Fixes #120

### CODE REVIEW

**Comment by smoya** [comment]
@jonaslagoni would you mind changing the PR title to what the main commit says: 

>feat!: split out definitions

File: package.json
Comment (smoya) [inline package.json:3]: Be mind about this.

**Comment by jonaslagoni** [comment]
@dalelane @derberg @fmvilas deciding on whether we should merge this will be better now rather than later (both I and @smoya need it for the next spec versions, or at least would be preferred). Anything you see as being blocked in order to merge this?

**Comment by derberg** [comment]
@jonaslagoni I'm ok for 3.0 release. I assume it is about splitting definitions. Just please provide more accurate title that will be used in commit and then in release notes. Also make it clear in description, what change is released and what breaking changes are actually expected

**Comment by jonaslagoni** [comment]
> @jonaslagoni I'm ok for 3.0 release. I assume it is about splitting definitions. Just please provide more accurate title that will be used in commit and then in release notes. Also make it clear in description, what change is released and what breaking changes are actually expected

Will do that 👍 

I am gonna resolve these conflicts as well, might require one additional PR to fix though.

**Comment by jonaslagoni** [comment]
PR is up that updates next-major with master https://github.com/asyncapi/spec-json-schemas/pull/187

File: package.json
Comment (jonaslagoni) [reply package.json:3]: Should be automatically taken care of.

**Comment by jonaslagoni** [comment]
Added a better description of the changes, let me know if anything else have to change 👍

**Comment by derberg** [comment]
@jonaslagoni looks perfect now, now you just need to solve conflicts

also, can you specify what is next plan, do we just update new 3.0 in parser just like that? no chances are needed? Did you try it already with the release candidate?

**Comment by jonaslagoni** [comment]
> also, can you specify what is next plan, do we just update new 3.0 in parser just like that? no chances are needed? Did you try it already with the release candidate?

Changes are needed but they are simple 🙂 I will make sure to update this PR as soon as it's released: https://github.com/asyncapi/parser-js/pull/423

**Comment by jonaslagoni** [comment]
> @jonaslagoni looks perfect now, now you just need to solve conflicts

I don't think I can solve those conflicts... I tried with #187 but it does not seem to apply them 🧐 And I do not have write access to the repo, so you will have to resolve them I think...

**Comment by jonaslagoni** [comment]
> Changes are needed but they are simple 🙂 I will make sure to update this PR as soon as it's released: [asyncapi/parser-js#423](https://github.com/asyncapi/parser-js/pull/423)

Or are they 🧐 Thought i did try to use the change, but apparently I did not. Looking into it.

**Comment by derberg** [comment]
you need to fork the repo, create a branch out from `next-major`, make sure it is up to date, also make sure your `master` is up to date. Them merge master to `next-major` in your fork, solve conflicts and then open a PR from your fork to upstream `next-major`. Makes sense?

**Comment by derberg** [comment]
please double-check, as the PR you link to doesn't show any changes to `package.json` while here we have conflicts with it

I recommend fresh fork, as something in your current fork might not be up to date with latest master, so you do not have last commit from Sergio from last week

**Comment by jonaslagoni** [comment]
> please double-check, as the PR you link to doesn't show any changes to `package.json` while here we have conflicts with it
> I recommend fresh fork, as something in your current fork might not be up to date with latest master, so you do not have last commit from Sergio from last week

I do not understand why that was not included in the other PR, cause I know I solved the conflicts there...

New PR to address that: https://github.com/asyncapi/spec-json-schemas/pull/188

**Comment by derberg** [comment]
conflicts still there 🤔 
your PR actually did not solve too many, like for example conflicting version of the package

**Comment by jonaslagoni** [comment]
> your PR actually did not solve too many, like for example conflicting version of the package

But I just kept "ours" in all cases, as I should right?

**Comment by jonaslagoni** [comment]
@derberg regarding parser and next steps, I am in contact with the JSON Schema community and trying to clarify where the problem is located and what to do about it. See https://json-schema.slack.com/archives/C5CF75URH/p1648137451821299

We are not quite sure where the fault lies, and what steps to take to resolve it. 

I am not sure whether we see it as a blocker for this?

**Comment by derberg** [comment]
oh that is a long thread 😅 

I'm not sure if you are referring to https://github.com/asyncapi/parser-js/pull/423 ? as it is draft, tests do not run, and tbh I'm not 100% sure what is the problem with the parser 😄

**Comment by smoya** [comment]
> @derberg regarding parser and next steps, I am in contact with the JSON Schema community and trying to clarify where the problem is located and what to do about it. See https://json-schema.slack.com/archives/C5CF75URH/p1648137451821299
> 
> We are not quite sure where the fault lies, and what steps to take to resolve it.
> 
> I am not sure whether we see it as a blocker for this?

The example you pasted in the thread fails the validation with https://json-schema.hyperjump.io. What is then the issue? AJV?

**Comment by jonaslagoni** [comment]
> The example you pasted in the thread fails the validation with https://json-schema.hyperjump.io. Where is then the issue? AJV?

With AJV, that example would be valid, that is the whole problem 😄

**Comment by jonaslagoni** [comment]
That being said, I dont think we should wait with merging this one, until I have fixed the parser 🤷 Cause it's not dependent IMO.

**Comment by jonaslagoni** [comment]
Here are the related issues: 
Ajv: https://github.com/ajv-validator/ajv/issues/1940
The bundling library: https://github.com/hyperjump-io/json-schema-bundle/issues/7

The next steps I think is to upgrade Ajv (related issue https://github.com/asyncapi/parser-js/issues/506)

**Comment by smoya** [comment]
> That being said, I dont think we should wait with merging this one, until I have fixed the parser 🤷 Cause it's not dependent IMO.

Indeed. The bug is in the parser but if we can't fix that it means we would never be able to use the next split out definitions right?

**Comment by jonaslagoni** [comment]
> Indeed. The bug is in the parser (in AJV dependency technically) but if we can't fix that it means we would never be able to use the next split out definitions right?

If the case was we would never be able to fix the bug, sure we would never be able to use the split definitions. However, that is not the case 🙂

**Comment by smoya** [comment]
> > Indeed. The bug is in the parser (in AJV dependency technically) but if we can't fix that it means we would never be able to use the next split out definitions right?
> 
> If the case was we would never be able to fix the bug, sure we would never be able to use the split definitions. However, that is not the case 🙂

My opinion is that I would **not merge** this until those issues have been solved. In consequence, **2.4** spec release will be crafted on top of the current (not split out) files.

Let's see what others say cc @derberg

**Comment by derberg** [comment]
@jonaslagoni did you try what you talked about on `spec 3.0` meeting?

**Comment by jonaslagoni** [comment]
> @jonaslagoni did you try what you talked about on `spec 3.0` meeting?

Yep, clarified it here: https://github.com/asyncapi/parser-js/pull/423

**Comment by jonaslagoni** [comment]
> @jonaslagoni did you try what you talked about on `spec 3.0` meeting?

Should we write some description of the problem encountered with the new bundled documents for validation libraries? If so, where? 🤔

**Comment by derberg** [comment]
@jonaslagoni just to clarify as I'm not sure why @smoya would not like to merge this one:
- if we merge it, it can cause only problems for people using the package directly? but this is what we do major release, so?
- thanks to https://github.com/asyncapi/parser-js/pull/423 parser will be all good?

**Comment by smoya** [comment]
> @jonaslagoni just to clarify as I'm not sure why @smoya would not like to merge this one:

Because by then (when I wrote my comment), there was uncertainty about the solution. However, 6 days ago https://github.com/asyncapi/parser-js/pull/423 added the solution for:

> One important note is, that with the new @asyncapi/specs package, the schema for validating JSON Schema's has to be removed in validation libraries

So I think if there is nothing else we should fix in our libraries, I'm ok **now** to move forward.

**Comment by derberg** [comment]
ok, cool

@jonaslagoni ping me for a coffee on Slack, we need to get on a call together as I want to help with solving merge conflicts and need you to confirm changes 😄

**Review by derberg — DISMISSED** [review]

**Comment by Dzixxx** [comment]
@jonaslagoni @derberg any updates on this topic? I cannot wait to check it out as a pre-release 🚀  😂

**Comment by derberg** [comment]
@jonaslagoni did you wait for me to merge? cause we have conflicts there again 😅

**Comment by derberg** [comment]
@jonaslagoni ok, they were not that scary like last time, done

**Comment by jonaslagoni** [comment]
@derberg I thought you where waiting for other codeowners, so didnt want to merge it 😃

Ready to merge then?

**Comment by derberg** [comment]
@jonaslagoni well 😄 tbh I thought they already did agree with this change. I would sware I saw they 👍🏼 somewhere 😅 

@fmvilas @dalelane just to be sure, can you folks approve 3.0 release that merge of this PR will trigger?

**Comment by smoya** [comment]
> @jonaslagoni well 😄 tbh I thought they already did agree with this change. I would sware I saw they 👍🏼 somewhere 😅

Maybe on the original PR https://github.com/asyncapi/spec-json-schemas/pull/128.

**Comment by fmvilas** [comment]
There are Markdown changes that I don't think belong here. Maybe you need to rebase or merge master?

**Comment by jonaslagoni** [comment]
> There are Markdown changes that I don't think belong here. Maybe you need to rebase or merge master?

Wonder where those changes come from 🧐

Regardless, they are automatically overwritten in the next update from the bot. Or do you want me to create a PR changing them back into what it is supposed to be?

**Comment by derberg** [comment]
sorry but what markdown changes are we talking about?

**Comment by jonaslagoni** [comment]
> sorry but what markdown changes are we talking about?

The changes to: `.github/workflows/lint-pr-title.yml`

**Comment by derberg** [comment]
oh, that makes sense, the original file has no spaces https://github.com/asyncapi/.github/blob/master/.github/workflows/lint-pr-title.yml. This is pretty stable workflow so do not expect it so change anytime soon, so no autoupdate in near future. But yeah, since it is just space in a comment 🤷🏼 😄 

I'm more concerned to merge it before 2.4
There is still some work to do and conflicts it can generate can make @smoya bald...... 😉

**Comment by jonaslagoni** [comment]
> There is still some work to do and conflicts it can generate can make @smoya bald...... 😉

Haha, I would not worry, I had a script that automatically split out the old files into definitions, I can resurface that somewhere and help @smoya split the final 2.4 schema 🙂

**Review by dalelane — APPROVED** [review]

**Comment by smoya** [comment]
All 2.4 changes regarding spec-json-schemas have been merged already. Let's merge this now, and port changes on `2022-04-release` then? @jonaslagoni @derberg

**Comment by jonaslagoni** [comment]
> All 2.4 changes regarding spec-json-schemas have been merged already. Let's merge this now, and port changes on `2022-04-release` then? @jonaslagoni @derberg

Sounds good to me 👍

**Comment by derberg** [comment]
the release is this week, right? Thursday latest if you want my help

I personally do not recommend introducing such a big change right before the release, as then you need again review of people and we are not always available 🤷🏼 it makes more sense IMHO to first release new schema for new spec and then next week new version of schema after split.

**Comment by smoya** [comment]
> the release is this week, right? Thursday latest if you want my help
> 
> I personally do not recommend introducing such a big change right before the release, as then you need again review of people and we are not always available 🤷🏼 it makes more sense IMHO to first release new schema for new spec and then next week new version of schema after split.

Let's do this. No more back and forth sounds good to me 👍

**Comment by Dzixxx** [comment]
Merge! 😂  You don't even know how much I'm looking forward to these changes (I even started replacing by webpack.config.js your actual released version) :D

**Comment by smoya** [comment]
> Merge! 😂 You don't even know how much I'm looking forward to these changes (I even started replacing by webpack.config.js your actual released version) :D

As mentioned by @derberg, the release is gonna happen this week, so you won't need to wait so much more for having this PR merged and released. Does it sounds reasonable?

**Comment by Dzixxx** [comment]
Sounds great, srlsy! <3

**Comment by jonaslagoni** [comment]
Gonna go through the changes in 2.4 and create a pull request so we can finish this 🤞

**Comment by jonaslagoni** [comment]
PR is up: https://github.com/asyncapi/spec-json-schemas/pull/212

**Comment by jonaslagoni** [comment]
@derberg @fmvilas @dalelane
Blocker to finish this PR: https://github.com/asyncapi/spec-json-schemas/pull/212

**Comment by jonaslagoni** [comment]
The last conflict does not seem like I will be able to solve. 

But I dont think there is anything else that need to happen for this to be ready.

**Comment by derberg** [comment]
Any help needed? what last conflict you have in mind that can't be solved. We won't be able to merge without solving conflicts 😄

**Comment by jonaslagoni** [comment]
> Any help needed? what last conflict you have in mind that can't be solved. We won't be able to merge without solving conflicts 😄

I do yes, it's the same as last time, where my PRs dont solve the conflicts 🙃

**Review by derberg — APPROVED** [review]
I'm ok to merge and release major

@dalelane @fmvilas ?

**Review by dalelane — APPROVED** [review]

**Review by fmvilas — APPROVED** [review]
Yup 🚀

**Comment by jonaslagoni** [comment]
😮 

Is it happening? 😆

**Comment by jonaslagoni** [comment]
Here goes nothing

/rtm

**Comment by jonaslagoni** [comment]
Damn, when targeted branches are removed, remaining PRs are closed as well with no way to retarget them...

**Comment by derberg** [comment]
@jonaslagoni oh, bot removes it by default. I restored it

**Pull request merged**
